src/tsdb.c

```diff
@@ -604,4 +605,37 @@ int SeriesCalcRange(Series *series,
 
 timestamp_t CalcWindowStart(timestamp_t timestamp, size_t window) {
     return timestamp - (timestamp % window);
+}
+
+timestamp_t SkipExpiredSamples(Series *series, long long *skipped) {
+    *skipped = 0;
+    if (series->totalSamples == 0) {
+        return 0;
+    }
+
+    size_t i = 0;
+    Chunk_t *chunk;
+    bool rev = false;
+    Sample sample = { 0 };
+    ChunkFuncs *funcs = series->funcs;
+    timestamp_t minTimestamp =
+        series->retentionTime ? series->lastTimestamp - series->retentionTime : 0;
+
+    SeriesTrim(series);
```

Comment:
@ashtul why do we trim here?

Reply:
Usually, we trim only when allocating a new chunk to avoid checking the dictionary on every `TS.ADD` command.
So if we do not trim here, we may return a number of chunks that exist but some are completely expired but were not trim due to our internal optimization.

Reply:
@ashtul we cannot trim here because this should be a readonly operation, i think its better to use series iterator and count the number of skipped samples.
It will also mean that we will "bleed" the abstraction layer, but its a better tradeoff compared to modifying the series.

Reply:
This is very similar to `passive expire` in redis. If we attempt to access an expired key which redis hasn't deleted yet, it will delete it then.
From user's POV, nothing has changed those values "should" have been removed but we haven't done it for performance reasons.